File: turing_machine.py
X_B = {
  ('B', 's1'): ('X', 'R', 's2'),
  ('B', 's2'): ('B', 'L', 's3'),
  ('X', 's3'): ('B', 'R', 's4'),
  ('B', 's4'): ('B', 'L', 's1')
}

def simulate(instructions):
  '''
    Setup intial state
    loop:
      lookup next instruction
      apply that instruction
  '''

  tape = ['B', 'B']
  head = 0
  state = 's1'

  for i in range(8):  # Should be inifinte

    print(tape)

    # print(state.rjust(4) + ': ' + ''.join(tape))
    # print('      ' + ' ' * head + '^')

    key = (tape[head], state)
    new_symbol, head_direction, new_state = instructions[key] # Which is exactly what we need to move
    # Apply/Run command
    tape[head] = new_symbol
    head += 1 if head_direction == 'R' else -1
    state = new_state


simulate(X_B)

File: turing_machine.js
const X_B = {
  'B_s1': ['X', 'R', 's2'],
  'B_s2': ['B', 'L', 's3'],
  'X_s3': ['B', 'R', 's4'],
  'B_s4': ['B', 'L', 's1']
}

function simulate(instructions) {
  // Initial state
  let tape = ['B', 'B']
  let head = 0
  let state = 's1'

  for (let i = 0; i <= 8; i++) {
    console.log(`${state}: ${tape.join('')}`)
    console.log(`    ${' '.repeat(head)}^`)

    let key = `${tape[head]}_${state}`
    let [newSymbol, headDirection, newState] = instructions[key]
    tape[head] = newSymbol
    head += headDirection === 'R' ? 1 : -1
    state = newState
  }
}

simulate(X_B)
